File: youtube-links/src/clients/ApiClient.ts
import {
  MyForbiddenException,
  MyNotFoundException,
  MyServerException,
  MyUnauthorizedException,
  MyValidationException,
} from '@/shared/exceptions/CustomException';
import type {
  ErrorResponse,
  ForbiddenErrorResponse,
  NotFoundErrorResponse,
  ServerErrorResponse,
  UnauthorizedErrorResponse,
  ValidationErrorResponse,
} from '@/shared/exceptions/ErrorResponse';
import { ExceptionType } from '@/shared/exceptions/ExceptionType';
import type { AxiosInstance, AxiosResponse } from 'axios';
import axios, { AxiosError } from 'axios';

export interface IApiClient {
  getReturnAxiosResponse(url: string): Promise<AxiosResponse>;
  get<TResponse>(url: string): Promise<TResponse>;
  postWithoutResponse<TRequest>(url: string, tRequest: TRequest): Promise<void>;
  post<TRequest, TResponse>(url: string, tRequest: TRequest): Promise<TResponse>;
  postReturnAxiosResponse<TRequest>(url: string, tRequest: TRequest): Promise<AxiosResponse>;
  put<TRequest>(url: string, tRequest: TRequest): Promise<void>;
  putWithoutResponse(url: string): Promise<void>;
  delete(url: string): Promise<void>;
}

export const delay = (ms: number) => new Promise((res) => setTimeout(res, ms));

export interface ApiOptions {
  baseUrl: string;
  authScheme: string;
  languageHeader: string;
}

export const ApiOptions: ApiOptions = {
  baseUrl: 'http://localhost:5000/',
  authScheme: 'Bearer',
  languageHeader: 'Accept-Language',
};

export class ApiClient implements IApiClient {
  private readonly baseUrl: string = ApiOptions.baseUrl;
  private readonly client: AxiosInstance;
  private readonly authScheme: string = ApiOptions.authScheme;
  private readonly languageHeader: string = ApiOptions.languageHeader;

  constructor() {
    this.client = axios.create({
      baseURL: this.baseUrl,
    });
  }

  async getReturnAxiosResponse(url: string): Promise<AxiosResponse> {
    try {
      const response = this.client.get(`${this.baseUrl}${url}`);
      return response;
    } catch (error) {
      this.HandleErrors(error as AxiosError);
      return Promise.reject();
    }
  }

  async get<TResponse>(url: string): Promise<TResponse> {
    try {
      const response = await this.client.get<TResponse>(`${this.baseUrl}${url}`);
      return response.data;
    } catch (error) {
      this.HandleErrors(error as AxiosError);
      return Promise.reject();
    }
  }

  async postWithoutResponse<TRequest>(url: string, tRequest: TRequest): Promise<void> {
    try {
      await this.client.post(`${this.baseUrl}${url}`, tRequest);
    } catch (error) {
      this.HandleErrors(error as AxiosError);
      return Promise.reject();
    }
  }

  async post<TRequest, TResponse>(url: string, tRequest: TRequest): Promise<TResponse> {
    try {
      const response = await this.client.post(`${this.baseUrl}${url}`, tRequest);
      return response.data;
    } catch (error) {
      this.HandleErrors(error as AxiosError);
      return Promise.reject();
    }
  }

  async postReturnAxiosResponse<TRequest>(url: string, tRequest: TRequest): Promise<AxiosResponse> {
    try {
      return await this.client.post(`${this.baseUrl}${url}`, tRequest);
    } catch (error) {
      this.HandleErrors(error as AxiosError);
      return Promise.reject();
    }
  }

  async put<TRequest>(url: string, tRequest: TRequest): Promise<void> {
    try {
      await this.client.put<TRequest>(`${this.baseUrl}${url}`, tRequest);
    } catch (error) {
      this.HandleErrors(error as AxiosError);
      return Promise.reject();
    }
  }

  async putWithoutResponse(url: string): Promise<void> {
    try {
      await this.client.put(`${this.baseUrl}${url}`);
    } catch (error) {
      this.HandleErrors(error as AxiosError);
      return Promise.reject();
    }
  }

  async delete(url: string): Promise<void> {
    try {
      await this.client.delete(`${this.baseUrl}${url}`);
    } catch (error) {
      this.HandleErrors(error as AxiosError);
      return Promise.reject();
    }
  }

  HandleErrors(error: AxiosError) {
    let apiError = error.response?.data as ErrorResponse;

    switch (apiError.type) {
      case ExceptionType.Validation:
        const validationErrorResponse = apiError as ValidationErrorResponse;
        throw new MyValidationException(validationErrorResponse.errors);
      case ExceptionType.Unauthorized:
        const unauthorizedErrorResponse = apiError as UnauthorizedErrorResponse;
        throw new MyUnauthorizedException();
      case ExceptionType.Forbidden:
        const forbiddenErrorResponse = apiError as ForbiddenErrorResponse;
        throw new MyForbiddenException();
      case ExceptionType.NotFound:
        const notFoundErrorResponse = apiError as NotFoundErrorResponse;
        throw new MyNotFoundException();
      case ExceptionType.Server:
      default:
        const serverErrorResponse = apiError as ServerErrorResponse;
        throw new MyServerException();
    }
  }
}

export const apiClient = new ApiClient();

File: YoutubeLinks.Blazor/Clients/ApiClient.cs
using System.Globalization;
using System.Net.Http.Headers;
using System.Net.Http.Json;
using Microsoft.AspNetCore.Components.WebAssembly.Http;
using Newtonsoft.Json;
using YoutubeLinks.Blazor.Auth;
using YoutubeLinks.Shared.Exceptions;

namespace YoutubeLinks.Blazor.Clients;

public interface IApiClient
{
    Task<HttpResponseMessage> Get(string url);
    Task<TResponse> Get<TResponse>(string url);
    Task Post<TRequest>(string url, TRequest tRequest);
    Task<TResponse> Post<TRequest, TResponse>(string url, TRequest tRequest);
    Task<HttpResponseMessage> PostReturnHttpResponseMessage<TRequest>(string url, TRequest tRequest);
    Task Put<TRequest>(string url, TRequest tRequest);
    Task Put(string url);
    Task Delete(string url);
}

public class ApiClient(
    HttpClient client,
    IJwtProvider jwtProvider) : IApiClient
{
    private const string AuthScheme = "Bearer";
    private const string LanguageHeader = "Accept-Language";
    private readonly string _baseUrl = client.BaseAddress?.ToString();

    public async Task<HttpResponseMessage> Get(string url)
    {
        await AddHeaderValues();
        var response = await client.GetAsync($"{_baseUrl}{url}");

        if (!response.IsSuccessStatusCode)
        {
            await HandleErrors(response);
        }

        return response;
    }

    public async Task<TResponse> Get<TResponse>(string url)
    {
        await AddHeaderValues();
        var response = await client.GetAsync($"{_baseUrl}{url}");

        if (!response.IsSuccessStatusCode)
        {
            await HandleErrors(response);
        }

        var tResponse = await response.Content.ReadFromJsonAsync<TResponse>();
        return tResponse;
    }

    public async Task Post<TRequest>(string url, TRequest tRequest)
    {
        await AddHeaderValues();
        var response = await client.PostAsJsonAsync($"{_baseUrl}{url}", tRequest);

        if (!response.IsSuccessStatusCode)
        {
            await HandleErrors(response);
        }
    }

    public async Task<TResponse> Post<TRequest, TResponse>(string url, TRequest tRequest)
    {
        await AddHeaderValues();
        var response = await client.PostAsJsonAsync($"{_baseUrl}{url}", tRequest);

        if (!response.IsSuccessStatusCode)
        {
            await HandleErrors(response);
        }

        var tResponse = await response.Content.ReadFromJsonAsync<TResponse>();
        return tResponse;
    }

    public async Task<HttpResponseMessage> PostReturnHttpResponseMessage<TRequest>(string url, TRequest tRequest)
    {
        await AddHeaderValues();

        var request = new HttpRequestMessage(HttpMethod.Post, $"{_baseUrl}{url}");
        request.Content = JsonContent.Create(tRequest);

        request.SetBrowserResponseStreamingEnabled(true);

        var response = await client.SendAsync(request, HttpCompletionOption.ResponseHeadersRead);

        if (!response.IsSuccessStatusCode)
        {
            await HandleErrors(response);
        }

        return response;
    }

    public async Task Put<TRequest>(string url, TRequest tRequest)
    {
        await AddHeaderValues();
        var response = await client.PutAsJsonAsync($"{_baseUrl}{url}", tRequest);

        if (!response.IsSuccessStatusCode)
        {
            await HandleErrors(response);
        }
    }

    public async Task Put(string url)
    {
        await AddHeaderValues();
        var response = await client.PutAsync($"{_baseUrl}{url}", null);

        if (!response.IsSuccessStatusCode)
        {
            await HandleErrors(response);
        }
    }

    public async Task Delete(string url)
    {
        await AddHeaderValues();
        var response = await client.DeleteAsync($"{_baseUrl}{url}");

        if (!response.IsSuccessStatusCode)
        {
            await HandleErrors(response);
        }
    }

    private async Task AddHeaderValues()
    {
        var token = await jwtProvider.GetJwtDto();
        client.DefaultRequestHeaders.Authorization =
            token is not null ? new AuthenticationHeaderValue(AuthScheme, token.AccessToken) : null;

        var currentCultureName = CultureInfo.CurrentCulture.Name;
        client.DefaultRequestHeaders.Add(LanguageHeader, currentCultureName);
    }

    private static async Task HandleErrors(HttpResponseMessage response)
    {
        var error = await response.Content.ReadAsStringAsync();
        var tResponse = JsonConvert.DeserializeObject<ErrorResponse>(error);

        switch (tResponse.Type)
        {
            case ExceptionType.Validation:
                var validationErrorResponse = JsonConvert.DeserializeObject<ValidationErrorResponse>(error);
                throw new MyValidationException(validationErrorResponse.Errors);
            case ExceptionType.Unauthorized:
                JsonConvert.DeserializeObject<UnauthorizedErrorResponse>(error);
                throw new MyUnauthorizedException();
            case ExceptionType.Forbidden:
                JsonConvert.DeserializeObject<ForbiddenErrorResponse>(error);
                throw new MyForbiddenException();
            case ExceptionType.NotFound:
                JsonConvert.DeserializeObject<NotFoundErrorResponse>(error);
                throw new MyNotFoundException();
            case ExceptionType.Server:
            default:
                JsonConvert.DeserializeObject<ServerErrorResponse>(error);
                throw new MyServerException();
        }
    }
}